### ISSUE TITLE
About ddim_step_with_logprob

### ISSUE THREAD

**Issue by Doctor-James** [hu-zijing/B2-DiffuRL#2]
I'd like to ask about the ddim_step_with_logprob function. When time_steps = 0 (i.e., the final denoising step), the timestep becomes equal to prev_timestep, causing variance = 0, which results in log_prob being NaN. Is my understanding correct? How is this issue resolved?

**Comment by hu-zijing** [comment]
Thanks for your question! In the final denoising step, timestep is 1 (or some other positive integer if using a special sampler) instead of 0. More generally, in denoising step t, the model denoise from $x_t$ to get $x_ {t-1}$, rather than denoising from $x_ {t+1}$ to get $x_t$. This file was copied and modified from diffuser. If you are interested in the details, you can further refer to it :-).

**Comment by Doctor-James** [comment]
In run_sample.py, for example, in the following code
```
        pipeline.scheduler.set_timesteps(config.sample.num_steps, device=accelerator.device)
        ts = pipeline.scheduler.timesteps
        extra_step_kwargs = pipeline.prepare_extra_step_kwargs(None, config.sample.eta)

        latents = [[noise_latents1]]
        log_probs = [[]]

        for i, t in tqdm(
            enumerate(ts),
            desc="Timestep",
            position=3,
            leave=False,
            disable=not accelerator.is_local_main_process,
        ): 
```
I ran a quick test

```
scheduler = DDIMSchedule(
   num_train_timesteps=1000,
   beta_schedule="scaled_linear",
   prediction_type="sample",
)
scheduler.set_timesteps(50, device)

print(scheduler.timesteps)

tensor([980, 960, 940, 920, 900, 880, 860, 840, 820, 800, 780, 760, 740, 720,
        700, 680, 660, 640, 620, 600, 580, 560, 540, 520, 500, 480, 460, 440,
        420, 400, 380, 360, 340, 320, 300, 280, 260, 240, 220, 200, 180, 160,
        140, 120, 100,  80,  60,  40,  20,   0], device='cuda:0')
```

So ultimately, doesn't 't' ever reach 0? I'm not very familiar with this aspect . I really appreciate your explanation.

**Comment by hu-zijing** [comment]
The DDIM Scheduler has a parameter `steps_offset`, which is set to 0 by default. This is why your test reach 0 ultimately. You can check this in detail at https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_ddim.py. 

However, if you run some pre-trained diffusion models, such as SD 1.4, SD 2.1 and SDXL, you will find these models set `steps_offset` to 1 in their scheduler config files (https://huggingface.co/CompVis/stable-diffusion-v1-4/blob/main/scheduler/scheduler_config.json, https://huggingface.co/stabilityai/stable-diffusion-2-1/blob/main/scheduler/scheduler_config.json, https://huggingface.co/stabilityai/stable-diffusion-xl-base-1.0/blob/main/scheduler/scheduler_config.json). This is why sampling from them will not cause $t=0$. 

Hope this can help you!

**Comment by Doctor-James** [comment]
Thanks for your help!

**Comment by Doctor-James** [comment]
Hello, I have one more question that I hope you can help me with. According to our previous discussion, in the pipeline_with_logprob function, when generating at t=1—that is, generating 𝑥0—random noise is still being added. This seems a bit counterintuitive, doesn't it?

**Comment by hu-zijing** [comment]
See Line 132-133 in `ddim_with_logprobs.py`: 

~~~python
    variance = _get_variance(self, timestep, prev_timestep)
    std_dev_t = eta * variance ** (0.5)
~~~

The `std_dev_t` is zero when `timestep=1`, making the random noise added to `x_0` zero. 

> Function `_get_variance()` and https://github.com/huggingface/diffusers/blob/main/src/diffusers/schedulers/scheduling_ddim.py can help you have a better understanding.

**Comment by Doctor-James** [comment]
Thank you for your response. If the noise added to 𝑥0 is zero, then wouldn't it be the same as what we discussed earlier—that during the final denoising step, i.e., at time_step=1, the log-probability log𝑝 becomes NaN? This is because in that case, 
prev_sample =prev_sample_mean

**Comment by Doctor-James** [comment]
In fact, after a careful comparison, I found that there is a key difference between the implementation of DDIM_with_prob and that of DDIM in Diffuser. DDIM_with_prob have
```
prev_timestep = torch.clamp(prev_timestep, 0, self.config.num_train_timesteps - 1)
```
This causes the original DDIM prev_t to be negative when t = 1, and alpha_prod_t_prev = self.final_alpha_cumprod = 1, so no noise will be added.
However, in DDIM_with_prob, this line of code causes prev_t = 0 when t = 1, and alpha_prod_t_prev = self.alphas_cumprod[0] = 0.9999, so the noise is not zero, which makes it possible to compute the logp
```
    alpha_prod_t_prev = torch.where(
        prev_timestep.cpu() >= 0, self.alphas_cumprod.gather(0, prev_timestep.cpu()), self.final_alpha_cumprod
    )
```
This is quite confusing to me. Do you have any suggestions?

**Comment by hu-zijing** [comment]
Thanks for your helpful comments! The noise added to $x_0$ is indeed not zero. I misunderstood this because I didn't notice that the parameter `set_alpha_to_one` has been set to false in commonly used pre-trained models (i.e., [scheduler config of sdv1.4](https://huggingface.co/CompVis/stable-diffusion-v1-4/blob/main/scheduler/scheduler_config.json)), which results in `final_alpha_cumprod` being equal to `self.alphas_cumprod[0]` instead of 1. 

After figuring out this, it seems that when running pre-trained models with diffuser, `alpha_prod_t_prev = self.final_alpha_cumprod = self.alphas_cumprod[0]` (instead of 1), since `set_alpha_to_one` is false. As you mentioned, `self.alphas_cumprod[0]=0.9999`. This means that in these pre-trained models, when t=1, random noise is still added. But in the eyes of humans, the noise does not affect the image quality, perhaps because it is too slight? 

Thank you for introducing me to this interesting phenomenon that I did not notice before. Yet, as I am currently unable to reach some device, the above conclusions are all based on my reading of the code. If there is anything wrong, please do not hesitate to point it out!

**Comment by Doctor-James** [comment]
大佬，我还想跟你探讨一下，

**Comment by Doctor-James** [comment]
大佬，我还想跟你探讨一下，我将DDPO用在了轨迹预测领域。我理解这个相当于是把diffusion去噪过程用在了训练阶段。这样可以正常训练吗，这貌似与diffusion原始的加噪预测噪声不太一样。
具体的问题，我用去噪过程得到的最终轨迹与GT做RL, 发现不收敛。结果我试了一下用最终轨迹与GT做监督学习（MSE loss）也不收敛，这其实也合理，因为本身也没有diffusion model是这样训练的。你觉得为什么在RL上，这一套又能成功呢
